### PR TITLE
More "immutability" across several classes

### PR DESCRIPTION
**Pull request by dnkoutso**
Looks big but it does not include actual logical changes, mostly moving stuff to initializer and cleaning up docs

### CODE REVIEW

File: lib/cocoapods/target.rb [hunk 1]
```diff
@@ -102,23 +120,11 @@ def static_framework?
...
-    # @return [Hash{String=>Symbol}] A hash representing the user build
-    #         configurations where each key corresponds to the name of a
-    #         configuration and its value to its type (`:debug` or `:release`).
-    #
-    attr_accessor :user_build_configurations
-
     # @return [PBXNativeTarget] the target generated in the Pods project for
     #         this library.
     #
     attr_accessor :native_target
```
Comment (dnkoutso) [inline lib/cocoapods/target.rb]: this the last property that needs to be removed from here.

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* WIP Make PodTarget and AggregateTarget immutable  
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/target/aggregate_target.rb [hunk 2]
```diff
@@ -3,25 +3,73 @@ module Pod
   # of the single Pods. The client targets will then depend on this one.
   #
   class AggregateTarget < Target
+    # Product types where the product's frameworks must be embedded in a host target
```
Comment (dnkoutso) [inline lib/cocoapods/target/aggregate_target.rb]: those were just moved, no changes

File: lib/cocoapods/target/aggregate_target.rb [hunk 3]
```diff
@@ -3,25 +3,73 @@ module Pod
...
+    #
+    # @note   The configurations are generated by the {TargetInstaller} and
+    #         used by {UserProjectIntegrator} to check for any overridden
+    #         values.
+    #
+    attr_reader :xcconfigs
+
+    # @return [Array<PodTarget>] The dependencies for this target.
+    #
+    attr_accessor :pod_targets
```
Comment (dnkoutso) [inline lib/cocoapods/target/aggregate_target.rb]: this will also have to be in another PR

File: lib/cocoapods/target/pod_target.rb [hunk 4]
```diff
@@ -74,13 +77,10 @@ def scoped(cache = {})
         if cache[cache_key]
           cache[cache_key]
         else
-          target = PodTarget.new(specs, [target_definition], sandbox, target_definition.label)
+          target = PodTarget.new(sandbox, host_requires_frameworks, user_build_configurations, archs, specs, [target_definition], target_definition.label)
           target.file_accessors = file_accessors
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: these properties here are still mutable but will be fixed in another PR.

**Comment by dnkoutso** [comment]
dont think this should ship with 1.5.0 so I am marking it for 1.6.0.

File: lib/cocoapods/installer/analyzer.rb [hunk 5]
```diff
@@ -430,29 +430,28 @@ def generate_targets(resolver_specs_by_target)
...
-          target.user_target_uuids = []
-          target.user_build_configurations = target_definition.build_configurations || { 'Release' => :release, 'Debug' => :debug }
+          user_project = nil
+          client_root = config.installation_root.realpath
+          user_target_uuids = []
+          user_build_configurations = target_definition.build_configurations || { 'Release' => :release, 'Debug' => :debug }
+          archs = []
           if target_definition.platform && target_definition.platform.name == :osx
-            target.archs = '$(ARCHS_STANDARD_64_BIT)'
+            archs = ['$(ARCHS_STANDARD_64_BIT)']
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer.rb]: API designates this must be an array so now it is.

File: lib/cocoapods/target/aggregate_target.rb [hunk 6]
```diff
@@ -3,25 +3,73 @@ module Pod
...
+    # @param [Xcodeproj::Project] user_project @see #user_project
+    # @param [Array<String>] user_target_uuids @see #user_target_uuids
+    # @param [Array<PodTarget>] pod_targets @see #pod_targets
+    #
+    def initialize(sandbox, host_requires_frameworks, user_build_configurations, archs, target_definition, client_root, user_project, user_target_uuids, pod_targets)
+      super(sandbox, host_requires_frameworks, user_build_configurations, archs)
       raise "Can't initialize an AggregateTarget with an abstract TargetDefinition" if target_definition.abstract?
-      super()
       @target_definition = target_definition
-      @sandbox = sandbox
```
Comment (dnkoutso) [inline lib/cocoapods/target/aggregate_target.rb]: this is set in super class was not needed

File: lib/cocoapods/target/pod_target.rb [hunk 7]
```diff
@@ -38,21 +38,24 @@ class PodTarget < Target
...
+      super(sandbox, host_requires_frameworks, user_build_configurations, archs)
       raise "Can't initialize a PodTarget without specs!" if specs.nil? || specs.empty?
       raise "Can't initialize a PodTarget without TargetDefinition!" if target_definitions.nil? || target_definitions.empty?
       raise "Can't initialize a PodTarget with only abstract TargetDefinitions" if target_definitions.all?(&:abstract?)
       raise "Can't initialize a PodTarget with an empty string scope suffix!" if scope_suffix == ''
-      super()
       @specs = specs.dup.freeze
       @test_specs, @non_test_specs = @specs.partition(&:test_specification?)
       @target_definitions = target_definitions
-      @sandbox = sandbox
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: same, set in superclass

**Comment by dnkoutso** [comment]
Updated rubocop todo as part of this PR since it introduces two temp offenses for long parameter list in `pod_target.rb`. Those will be reduced as we further refactor/cleanup stuff.
